### PR TITLE
Setting default value even if that value is null

### PR DESCRIPTION
**Pull request by harishvashistha**
Hi Team,

We have a use case as mention in this issue https://github.com/networknt/json-schema-validator/issues/553 which is causing failures in downstream systems where "x":null has different meaning than non existance of "x". can you please evaluate if this change can go in future release else we would have to put some hacky code before calling json-schema-validator which modifies the json which is going to be validated and thereby breaks the purpose of this library.

Please help

### CODE REVIEW

**Comment by stevehu** [comment]
@harishvashistha I think this PR covers the previous one. Could we close that one and only leave this one open? Thanks.

**Comment by harishvashistha** [comment]
PR https://github.com/networknt/json-schema-validator/pull/554/files has changes related to file

`src/main/java/com/networknt/schema/utils/JsonNodeUtil.java`

so that nullable fields succeeds when handleNullable is true and null value supplied. This current PR does not have that change.
This PR is related to setting null as default value when `"default": "null"` is specified in schema definition.

Do you want me to merge these two PRs into this one for tracking?

**Comment by stevehu** [comment]
No necessary. You are using the same branch to track both PRs so they merged already automatically. If you look at this PR, there are three commits.

**Comment by harishvashistha** [comment]
@stevehu apologies for confusion, but this PR does not have code from another PR as i reverted that code manually while creating this branch, as I wanted both the PRs to show relevant changes only for the issue they are targeting and mentioned in subject. Please consider both the PRs with respect to the file changes they have. 
When can we expect these two PRs in official master as we are planning our project release to production soon?

**Comment by stevehu** [comment]
The changes look good to me. I just want other developers to have some input on the changes as they are more familiar with this feature. Once merged, we are expecting a release soon.

File: src/main/java/com/networknt/schema/PropertiesValidator.java
Comment (SiemelNaran) [inline src/main/java/com/networknt/schema/PropertiesValidator.java:141]: In ItemsValidator they have similar code
```
                if (defaultNode != null && defaultNode.isNull()) {
                    defaultNode = null;
                }
```
Maybe this should be removed?

Comment (SiemelNaran) [reply src/main/java/com/networknt/schema/PropertiesValidator.java:141]: Can you add unit test to JsonWalkApplyDefaultsTest?  You'll have to change walk-schema-default.json, adding `"default": null` somewhere, and similar changes to walk-data-default.json.

**Review by SiemelNaran — CHANGES_REQUESTED** [review]

File: src/main/java/com/networknt/schema/PropertiesValidator.java
Comment (harishvashistha) [reply src/main/java/com/networknt/schema/PropertiesValidator.java:140]: > In ItemsValidator they have similar code
> 
> ```
>                 if (defaultNode != null && defaultNode.isNull()) {
>                     defaultNode = null;
>                 }
> ```
> 
> Maybe this should be removed?

In ItemsValidator, this code is fine as Library specification says `applyArrayDefaults if true then apply defaults inside json arrays if the attribute is explicitly null` so code in here is checking if incoming item is null and default is not null then set that not null item in the array in place of null as there is no point of setting null in place of null. This is depicted in this code:

```
for (JsonNode n : arrayNode) {
                if (n.isNull() && defaultNode != null) {
                    arrayNode.set(i, defaultNode);
                    n = defaultNode;
                }
                doWalk(validationMessages, i, n, rootNode, at, shouldValidateSchema);
                i++;
            }
```

Comment (harishvashistha) [reply src/main/java/com/networknt/schema/PropertiesValidator.java:140]: > Can you add unit test to JsonWalkApplyDefaultsTest? You'll have to change walk-schema-default.json, adding `"default": null` somewhere, and similar changes to walk-data-default.json.

I have added some required test cases, Please have a look.

File: src/test/resources/schema/walk-schema-default.json
Comment (SiemelNaran) [inline src/test/resources/schema/walk-schema-default.json:27]: Do you need lines 19-27? It seems lines 14-18 is good enough.

File: src/main/java/com/networknt/schema/PropertiesValidator.java
Comment (SiemelNaran) [reply src/main/java/com/networknt/schema/PropertiesValidator.java:140]: > In ItemsValidator, this code is fine as Library specification says `applyArrayDefaults if true then apply defaults inside json arrays if the attribute is explicitly null` so code in here is checking if incoming item is null and default is not null then set that not null item in the array in place of null as there is no point of setting null in place of null. This is depicted in this code:

This makes sense.  However at some point I was thinking we may want to add an event listener, that's why I created the ApplyDefaultsStrategy class to add more stuff in the future if we want to.  Now we may want the listener to count the number of places where the default is missing or something else, so replacing null with null and firing the listener makes sense in that case.  What do you think?

File: src/main/java/com/networknt/schema/PropertiesValidator.java
Comment (harishvashistha) [reply src/main/java/com/networknt/schema/PropertiesValidator.java:144]: make sense, updating accordingly

File: src/test/resources/data/walk-data-default.json
Comment (harishvashistha) [reply src/test/resources/data/walk-data-default.json:10]: I applied this test case just to cover some if/else condition I have put in place. But I have removed it to be according to my description of changes

File: src/test/resources/schema/walk-schema-default.json
Comment (harishvashistha) [reply src/test/resources/schema/walk-schema-default.json:22]: @SiemelNaran I removed one unnecessary schema case. But other schema case I have left from 19-21 is covering code ranging from 136 to 138 in PropertiesValidator.java file. I think this case makes sense.

**Comment by harishvashistha** [comment]
@SiemelNaran I have fixed all the review comments. Please try to release this code with PR https://github.com/networknt/json-schema-validator/pull/554 as we urgently need these fixes for our Tuesday release next week in production.

File: src/main/java/com/networknt/schema/ItemsValidator.java
Comment (SiemelNaran) [inline src/main/java/com/networknt/schema/ItemsValidator.java:128]: This part was an open question.  If you are removing these lines, then it means you are applying default even if it is null in the case of a json array.  So then for parallel it make sense to do this for the other file where the
```
            boolean applyDefault = propertyNode == null
                    || (applyDefaultsStrategy.shouldApplyPropertyDefaultsIfNull() && propertyNode.isNull() && !defaultNode.isNull());
```
should be changed to
```
            boolean applyDefault = propertyNode == null || propertyNode.isNull()  && applyDefaultsStrategy.shouldApplyPropertyDefaultsIfNull();
```

On the other hand if you want to skip applying default if the default is null node, then keep line lines the other fille as is but bring back lines 126-128. 

Will let you decide how to want to go, but keep things in parallel.

**Review by SiemelNaran — APPROVED** [review]
Approved after minor comment addressed.  I'll review again afterwards.

File: src/main/java/com/networknt/schema/ItemsValidator.java
Comment (harishvashistha) [reply src/main/java/com/networknt/schema/ItemsValidator.java:128]: @SiemelNaran I have aligned the logic at both the places, by setting null even if value provided is null, because of your listener logic that may come in future for some requirement.

**Comment by harishvashistha** [comment]
@SiemelNaran @stevehu @prashanthjos can we get these changes in master release please. This PR is urgently required with PR https://github.com/networknt/json-schema-validator/pull/554 as we are planning a relase on Tuesday.
I hope you would understand the urgency for these changes.

**Comment by stevehu** [comment]
@harishvashistha I am not familiar with the feature and depend on other team members to review the changes. Once the PR is reviewed and merged, we will have a release.

**Comment by harishvashistha** [comment]
@stevehu thanks for the update, may be if I can ask @SiemelNaran  and @prashanthjos if you guys can help me with these two PRs, it would be a great gesture on your side.
I am fully dependent on this library to move to production with our project. Please help with these PRs: https://github.com/networknt/json-schema-validator/pull/554 , https://github.com/networknt/json-schema-validator/pull/555

**Review by SiemelNaran — APPROVED** [review]

**Comment by SiemelNaran** [comment]
This PR looks good to me.  I don't know if I have the ability to merge to master.  I will try now.

**Review by stevehu — APPROVED** [review]

**Pull request merged**
